File: src/inbox/approve.ts
// Approving a review hand-back from the Inbox = moving the agent's Under
// Review card(s) to the board's Done column. Shared by the viewer's
// "Approve → Done" strip and the `e` key on an agent row, so both surfaces
// mean the same thing.

import type { ReviewHandback } from '@/inbox/route'

export type CardMover = (project: string, query: string, toColumn: string) => Promise<void>

export interface ApproveOutcome {
  /** Hand-backs actually moved to Done. */
  moved: ReviewHandback[]
  /** Hand-backs whose board has no Done column — nothing to move them to. */
  skipped: ReviewHandback[]
  /** First failure, if a move threw; later hand-backs are left untouched. */
  error?: { handback: ReviewHandback; cause: unknown }
}

/** Move every approvable hand-back to Done, in order, stopping at the first
 *  failure (a half-approved batch is visible on the board; a silently
 *  swallowed error is not). */
export async function approveHandbacks(handbacks: ReadonlyArray<ReviewHandback>, move: CardMover): Promise<ApproveOutcome> {
  const out: ApproveOutcome = { moved: [], skipped: [] }
  for (const h of handbacks) {
    if (!h.doneColumn) { out.skipped.push(h); continue }
    try {
      await move(h.project, h.query, h.doneColumn)
      out.moved.push(h)
    } catch (cause) {
      out.error = { handback: h, cause }
      break
    }
  }
  return out
}

/** Hub errors arrive as a JSON `{error}` body inside the thrown message. */
export function hubErrorText(e: unknown): string {
  const m = e instanceof Error ? e.message : String(e)
  try {
    const j = JSON.parse(m) as { error?: unknown }
    if (typeof j?.error === 'string') return j.error
  } catch { /* not JSON */ }
  return m
}

File: android/app/src/main/kotlin/io/amar/console/data/inbox/Approve.kt
package io.amar.console.data.inbox

import io.amar.console.data.spaces.SpacesRepository.ReviewHandback

// Approving a review hand-back from the Inbox = moving the agent's Under
// Review card(s) to the board's Done column. Shared by the agent screen's
// "Approve → Done" strip and swipe-done on an agent row, so both surfaces
// mean the same thing (port of src/inbox/approve.ts).

data class ApproveOutcome(
    /** Hand-backs actually moved to Done. */
    val moved: List<ReviewHandback>,
    /** Hand-backs whose board has no Done column — nothing to move them to. */
    val skipped: List<ReviewHandback>,
    /** First failure, if a move failed; later hand-backs are left untouched. */
    val failed: ReviewHandback? = null,
)

/** Move every approvable hand-back to Done, in order, stopping at the first
 *  failure (a half-approved batch is visible on the board; a silently
 *  swallowed error is not). [move] returns false on failure. */
suspend fun approveHandbacks(
    handbacks: List<ReviewHandback>,
    move: suspend (project: String, query: String, toColumn: String) -> Boolean,
): ApproveOutcome {
    val moved = ArrayList<ReviewHandback>()
    val skipped = ArrayList<ReviewHandback>()
    for (h in handbacks) {
        val col = h.doneColumn
        if (col == null) { skipped += h; continue }
        if (move(h.project, h.query, col)) moved += h
        else return ApproveOutcome(moved, skipped, failed = h)
    }
    return ApproveOutcome(moved, skipped)
}
